---
#output: github_document
output: html_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# PlasmoSim

<!-- badges: start -->
[![develop checks](https://github.com/mrc-ide/PlasmoSim/workflows/checks_develop/badge.svg)](https://github.com/mrc-ide/PlasmoSim/actions)
[![master checks](https://github.com/mrc-ide/PlasmoSim/workflows/checks_master/badge.svg)](https://github.com/mrc-ide/PlasmoSim/actions)
<!-- badges: end -->

PlasmoSim is an R package for simulating both epidemiological and genetic data from a simple model of Plasmodium falciparum transmission. The model itself is fairly lightweight, which speeds up computation times and in turn increases the number of demes (partially isolated sub-populations) that we can simulate. This allows us to explore interesting questions relating to spatial patterns of P. falciparum genetic structure.

## Installation

You can install the development version of PlasmoSim from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mrc-ide/PlasmoSim")
```

Now we can load the Plasmosim package, along with others that will be useful:

```{r, message=FALSE}
# load packages
library(PlasmoSim)
library(ggplot2)
library(tidyverse)
library(kableExtra)
```

## Example

```{r, echo=FALSE}
set.seed(3)
```

PlasmoSim returns two types of output:

1. Population-level data giving basic epidemiological measures (e.g. prevalence) on each day of simulation
2. Individual-level data, including parasite genotypes found in the host at specific sampling points

For the latter, we need to define our desired sampling strategy using a data.frame. This specifies the demes that we will sample from, at what time points, and how many hosts will be drawn at random from the population:

```{r}
# define individual-level sampling via a data.frame
sample_df <- data.frame(deme = 1,
                        time = 365,
                        n = 100)
```

Now we run the main simulation function:

```{r}
# run simulation
sim1 <- sim_falciparum(H = 1000,                     # human population size
                       M = 5000,                     # adult female mosquito population size
                       seed_infections = 100,        # number of infected hosts at time 0
                       L = 24,                       # number of loci
                       sample_dataframe = sample_df  # sampling data.frame
                       )
```

Daily output is stored in long format, which makes it easy to produce plots:

```{r, out.width="70%", out.height="70%"}
# basic plot of prevalence in "I" state (infected)
sim1$daily_values |>
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = 100 * I / 1000)) +
  ylab("Prevalence (%)")
```

We can see that prevalence jumped to 10% (100 infected hosts in the population of 1000) on day 13. This is because we set the simulation going with 100 seed infections, which here means new liver-stage infections. The default time from liver-stage to blood-stage (the intrinsic incubation period) is set to \code{u = 12} by default, hence these all emerging together on day 13. After this point we can see dynamic and stochastic changes in the number infected.

Individual-level output is also returned in long format. We will use the `kable` package to make this slightly easier to read:

```{r}
# take a peek at basic individual-level output, without the haplotypes column
sim1$indlevel |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) |>
  scroll_box(width = "1000px", height = "400px")
```

The first few columns tell us when and where (i.e. which deme) sampling occurred, the ID of the host, and whether they were positive for malaria parasites. The next two columns give genetic data (positive samples only). The `haplotypes` column gives the raw information at all 24 loci. Note that each host can be infected with multiple strains, meaning this element is actually a matrix with one row for each strain and one column for each locus. We can see this by printing out the full element for a malaria-positive host:

```{r}
# print raw genetic data
sim1$indlevel |>
  filter(sample_ID == 394) |>
  pull(haplotypes)
```

Returning to the table, the final column gives the `haplo_ID`. This is a hash of the information contained in each row of the `haplotypes` matrix, meaning each unique combination of values over all loci will be given a unique name. This can be very useful when we only care about unique genotypes and not the locus-by-locus information contained in those genotypes. For example, for the same individual as before:

```{r}
# print haplo_ID
sim1$indlevel |>
  filter(sample_ID == 394) |>
  pull(haplo_ID)
```

But what do the values in the `haplotypes` matrix actually mean? Although we have described them as haplotypes, they **do not** (yet) represent genetic information. Instead, each value specifies the ancestor that the information is descended from at the start of the simulation. For example, if we see a value 68 then we know that, at this locus, the information eventually traces back to the 68th seed infection at the start of the simulation. There are two reasons for encoding information like this:

1. It allows us to ask questions about identify by descent (IBD), and not just identity by state (IBS). If two samples have the same value at the same locus then we know they are descended from a common ancestor some time between the start of the simulation and the present day.
2. We can always go from this ancestral encoding to genotypes; we simply have to define a genetic value for each unique ancestor at each locus (we will see an example of this shortly). However, the converse is not true; we cannot always go back from genotypes to ancestry.

Note that two samples having the value 68 does not imply that they are both *direct* descendants of the 68th seed infection. Rather, it implies that these samples have a common ancestor some time between the start of the simulation and the present day, and *this common ancestor* is descended from the 68th seed infection. It is much more likely that the common ancestor is much more recent than going all the way back to the start of the simulation.

Look again at the `haplotypes` matrix for this poly-clonally infected host. Notice that each row contains blocks of the values 68 and 42:

```{r}
# print raw genetic data
sim1$indlevel |>
  filter(sample_ID == 394) |>
  pull(haplotypes)
```

What we are looking at here is recombination break points. We could use this matrix to identify blocks of IBD within a sample, or compare this matrix with another to explore IBD between samples.

Let's see how we can convert these ancestral encodings to genotypes.

## Example with migration

PlasmoSim incorporates migration of hosts between demes using a migration matrix. We have to be careful when simulating stochastic movement of people between demes, as if we simply give each person a random chance of moving from one deme to another taken from this migration matrix then deme population sizes will drift up and down at random and eventually we will see demes emptying entirely. To get around this we do matched migration, meaning people swap places between demes rather than moving independently. This satisfies the migration matrix, while also ensuring that population sizes stay constant throughout the simulation. We do not model mosquito movement in this version of the package.

We start by creating a grid of demes, and for each deme specifying its properties, such as the mosquito population size and the number of seeding infections. In our case we will make the density of mosquitoes (and hence the EIR) increase from left to right of the domain. Note that human population sizes must be the same for all demes due to the migration constraint described above. We will start by seeding infections in just one deme in the lower left corner:

```{r}
# define x- and y-locations of demes
demes_x <- seq(0, 100, 10)
demes_y <- seq(0, 100, 10)

# get coordinates into dataframe
deme_df <- expand.grid(demes_x, demes_y)
names(deme_df) <- c("x", "y")

# add deme-specific properties
deme_df <- deme_df %>%
  dplyr::mutate(deme = seq_len(nrow(deme_df)),
                M = 500 + x, # mosquito population size increasing from left to right
                seed_infections = ifelse(x == 0 & y == 0, 50, 0))

head(deme_df)
```

Next we need to make a migration matrix. We first calculate the distance between all demes, and then we will create a small amount of migration between adjacent demes only. The migration matrix must sum to one over rows to be accepted by the program:

```{r}
# define migration matrix based on distance
d <- deme_df %>%
  dplyr::select(x, y) %>%
  dist() %>%
  as.matrix()
mig_matrix <- 0.01 * (d <= 10)

# ensure migration probabilities sum to 1 over rows
diag(mig_matrix) <- 0
diag(mig_matrix) <- 1 - rowSums(mig_matrix)
```

Next we need to define our sampling strategy. We will sample 100 individuals from a subset of demes after a full 10 years of simulation (genetic patterns take a long time to settle down, much longer than prevalence for example). For convenience we produce the sampling dataframe directly from the `deme_df` dataframe defined above, although only the columns `deme`, `time` and `n` will be used sampling.

```{r}
# define output dataframe
sample_df <- deme_df %>%
  dplyr::filter(deme %in% c(13, 26, 35, 109)) %>%
  dplyr::mutate(time = 30 * 365,
                n = 100)

sample_df
```

Now we can run the simulation, using the values in `deme_df` to specify the mosquito population size and the seeding infections:

```{r}
# simulate
set.seed(1)
sim_mig <- sim_falciparum(H = 100,
                          M = deme_df$M,
                          seed_infections = deme_df$seed_infections,
                          mig_matrix = mig_matrix,
                          L = 24,
                          sample_dataframe = sample_df)
```

As before, we can produce a simple plot of prevalence over time, now broken down by deme. We can see how the wave of infection swept through the demes as infected hosts migrated through the space:

```{r, out.width="70%", out.height="70%"}
# basic plot of prevalence in "I" state (infected)
sim_mig$daily_values %>%
  dplyr::filter(time < 365 * 10) %>%
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = 100 * I / 100, col = as.factor(deme)), show.legend = FALSE) +
  ylab("Prevalence (%)")
```

We can visualise this spatially by filtering to a certain timepoint and then producing a raster plot. For example, here is the prevalence map 5 years into the simulation:

```{r, out.width="70%", out.height="70%"}
# subset daily output to specific timepoint and merge back with deme properties
daily_values_sub <- sim_mig$daily_values %>%
  dplyr::filter(time == 5*365) %>%
  dplyr::left_join(deme_df, by = "deme")

# produce raster plot
daily_values_sub %>%
  ggplot() + theme_bw() +
  geom_raster(aes(x = x, y = y, fill = I)) +
  scale_fill_viridis_c(option = "magma", name = "Prevalence (%)")
```

We can see that five years into our simulation the wave of infection had still not quite reached the top-right corner of our domain. Prevalence also appears to be higher towards the right of the domain, a consequence of assuming increasing mosquito population size from left to right. We can produce the same plot at 30 years into simulation:

```{r, out.width="70%", out.height="70%"}
# subset daily output to specific timepoint and merge back with deme properties
daily_values_sub <- sim_mig$daily_values %>%
  dplyr::filter(time == 30*365) %>%
  dplyr::left_join(deme_df, by = "deme")

# produce raster plot
daily_values_sub %>%
  ggplot() + theme_bw() +
  geom_raster(aes(x = x, y = y, fill = I)) +
  scale_fill_viridis_c(option = "magma", name = "Prevalence (%)")
```

Now prevalence appears to be roughly at equilibrium, with higher transmission uniformly along the right of the domain.

Returning to our genetic sample, we asked the program to draw 100 individuals at random from four demes on the final day of simulation. We can calculate the pairwise genetic identity between all pairs of individuals in this sample using the \code{get_identity_matrix()} function. Genetic identity is calculated as the proportion of identical genetic values over all possible pairs of haplotypes compared between two individuals. This means two individuals will only have identity = 1 if they are monoclonal and matching at every locus. It follows that an individual will be less than perfectly identical to itself if it carries multiple distinct haplotypes. As long as genetic values represent ancestry, as they do in this simulation, what we are really measuring here is average identity by descent (IBD) between haplotypes. If genetic values were converted into alleles prior to running this function then we would be measuring identity by state (IBS).

```{r, out.width="70%", out.height="70%"}
# get pairwise genetic identity between all individuals
indlevel_identity <- get_identity_matrix(sim_mig)

# get into long form dataframe
n_pos <- nrow(indlevel_identity[[1]])
df_indlevel_identity <- expand.grid(1:n_pos, 1:n_pos)
df_indlevel_identity$value <- as.vector(indlevel_identity[[1]])

# plot pairwise matrix
ggplot(df_indlevel_identity) + theme_void() +
  geom_raster(aes(x = Var1, y = Var2, fill = value)) +
  scale_fill_viridis_c(option = "magma", name = "Identity by descent")
```

We can also use the same function to summarise genetic identity at the deme level rather than the individual level through the `deme_level = TRUE` argument:

```{r, out.width="40%", out.height="40%", fig.width=4, fig.height=2.5}
# get pairwise genetic identity between all demes
demelevel_identity <- get_identity_matrix(sim_mig, deme_level = TRUE)

# get into long form dataframe
n_deme <- nrow(demelevel_identity[[1]])
df_demelevel_identity <- expand.grid(deme1 = sample_df$deme, deme2 = sample_df$deme)
df_demelevel_identity$IBD <- as.vector(demelevel_identity[[1]])

# plot pairwise matrix
ggplot(df_demelevel_identity) + theme_bw() +
  geom_raster(aes(x = as.factor(deme1), y = as.factor(deme2), fill = IBD)) +
  xlab("deme1") + ylab("deme2") +
  scale_fill_viridis_c(option = "magma", name = "Identity by descent")
```

We can see that some demes are more closely related than others. We can visualise this spatially by drawing edges between our sampled demes, with edge thickness/colour proportional to IBD:

```{r, out.width="70%", out.height="70%"}
# append deme coordinates
sample_df_simple <- sample_df %>%
  dplyr::select(x, y, deme)

df_demelevel_identity <- df_demelevel_identity %>%
  dplyr::rename(deme = deme1) %>%
  dplyr::left_join(sample_df_simple, by = "deme") %>%
  dplyr::rename(deme1 = deme,
                deme = deme2) %>%
  dplyr::left_join(sample_df_simple, by = "deme") %>%
  dplyr::rename(deme2 = deme)

# plot pairwise IBD
df_demelevel_identity %>%
  dplyr::filter(deme1 != deme2) %>%
  ggplot() + theme_bw() +
  geom_segment(aes(x = x.x, y = y.x, xend = x.y, yend = y.y, size = IBD, col = IBD),
               lineend = "round") +
  xlim(c(0, 100)) + ylim(c(0, 100)) +
  xlab("x") + ylab("y") +
  guides(size = "none")
```

On average we would expect to see thaht demes closer together would be more highly related, although this will also be influenced in complex ways by relative transmission intensity.
